hail/python/hail/matrixtable.py

```diff
@@ -2592,9 +2593,13 @@ def _select_all(self,
         base, cleanup = self._process_joins(*all_exprs)
         jmt = base._jvds
 
+        if row_key is not None:
+            jmt = jmt.keyRowsBy([])
```

Comment:
I tried to make `_select_all` be like the other select methods—not allowed to change keys—but I couldn't figure out how. It's only used in two places, `export_gen` and `export_plink`. Neither one changes row keys, so I could remove the `row_key` parameter and remove the confusion, but they both change column keys (to `[]`), and I couldn't figure out how to factor that out. If I change column keys before the `_select_all`, then all the expressions are now on the wrong MatrixTable, and if I do it after, the `selectCols` step might be changing or removing column keys.

So I decided to keep the `row_key` parameter for symmetry. I'm not entirely happy with it, and I'm open to suggestions, but I don't think this is too bad. We only call `keyRowsBy` when `row_key is not None`, which means the caller is asking to change the row key. If we're changing the key, we expect to have to do a scan and shuffle. The only case where the generated IR might cause extra scans is when the new key is a non-empty prefix of the old key. That case seems unlikely to come up in `select_all`, and anyways I think, rather than complicating the python code, the right fix is to make the optimizer simplify this to the ideal IR.

Reply:
ok.